### ISSUE TITLE
Flags out of scope for the command are ignored via --config

### ISSUE THREAD

**Issue by justinhorvitz** [bazelbuild/bazel#22980]
### Description of the bug:

```
$ bazel build --verbose_test_summary
ERROR: --verbose_test_summary :: Unrecognized option: --verbose_test_summary
```

This is expected because `--verbose_test_summary` is only for the `test` command and not `build`. However, the following succeeds:

```
$ echo "build:abc --verbose_test_summary" >> .bazelrc
$ bazel build --config=abc
```

**Comment by justinhorvitz** [comment]
@fmeum might this be related to your work on `common` (https://github.com/bazelbuild/bazel/pull/18130)?

**Comment by fmeum** [comment]
That seems likely indeed. I'm OOO at the moment, but could look into it in ~2 weeks.

**Comment by fmeum** [comment]
I looked into this and confirm that my change causes this: In order to support ignoring `common` arguments in flag expansions, `fallbackData` is passed into the expansion function regardless of which commands the expanded arguments were specified on. As a result, all commands in config expansions behave like `common`.

The most straightforward fix I can think of is to annotate each individual arg with the `fallbackData` that applies to it in arguments to the `parse` functions in `OptionsParser`. I initially tried to make very changes and just run the `parse` function for each `common` or non-`common` run of args, but that resulted in issues with option priorities that I ultimately couldn't resolve since `OptionsParserImplResult`s can't easily be merged.

I now think that the main `parse` function has to be modified to accept something like `List<Pair<OptionsData, List<String>>` as `args`. @justinhorvitz What do you think about that?

**Comment by justinhorvitz** [comment]
I looked into this too and see the complications. By "main `parse` function" do you mean [this](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/common/options/OptionsParserImpl.java;l=486;drc=0803ad3770155be411a703950502dc71743017bd)?

I think what you said about `List<Pair<OptionsData, List<String>>` is on the right track, but maybe it should be `List<Pair<String, OptionsData>>` since common and non-common args can be intermingled due to recursive config expansion. And we can make the type easier to understand by using a `record ArgAndFallbackData(String arg, OptionsData fallbackData)`.

In the non-config case where all args get the same fallback data, we can use a lazy `Lists.transform(args, arg -> new ArgAndFallbackData(arg, fallbackData))`.

There are some potential complications with that `argsPreProcessor` that I didn't think through.

**Comment by fmeum** [comment]
I started out with `List<Pair<String, OptionsData>>` outside the function you linked and ran into issues when parsing a `--copt -DFOO` arg-by-arg, which doesn't work. I agree that with the logic moved into `parse`, that representation is better.

**Comment by iancha1992** [comment]
A fix for this issue has been included in [Bazel 7.4.0 RC1](https://github.com/bazelbuild/bazel/releases/tag/7.4.0rc1). Please test out the release candidate and report any issues as soon as possible.
If you're using Bazelisk, you can point to the latest RC by setting `USE_BAZEL_VERSION=7.4.0rc1`. Thanks!
